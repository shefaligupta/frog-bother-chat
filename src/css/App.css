.app {
  padding: 10px;
}

.app-header {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.app-title {
  font-weight: 400;
}

.button-henchmen {
  margin-bottom: 5px;
}

.select-henchman, .active-button {
  width: 100%;
}

.chat-window {
  display: flex;
  height: 300px;
  flex: 1;
}

.henchman {
  display: flex;
  flex-direction: column;
  flex: 0.2;
  padding-right: 20px;
}

.chat-history {
  display: flex;
  flex: 1;
  border: 1px solid #eee;
  flex-direction: column;
  padding: 10px;
  overflow: auto;
}

.write-message {
  display: flex;
  margin-top: 20px;
  justify-content: flex-end;
}

.message {
  margin-bottom: 10px;
}

.message-time {
  font-size: 13px;
}

.type-message {
  flex: 0.7;
  margin-right: 20px !important;
}

.send-message {
  flex: 0.1;
}

.active-button {
  border: 1px solid #000 !important;
}