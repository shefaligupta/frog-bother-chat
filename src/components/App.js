import React, {useRef, useState} from 'react';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import '../css/App.css';
import HenchmenList from './HenchmenList';
import ChatHistory from './ChatHistory';

/**
 * App Class
 * @returns {*}
 * @constructor
 */
const App = () => {
  const [message, setMessage] = useState('');
  const [chat, setChat] = useState(new Map());
  const [henchman, setHenchman] = useState('Sonny');

  const inputEl = useRef(null);

  /**
   * sends the message i.e. calls the setter for chatHistory map
   * which in turn runs the UI update.
   */
  const sendMessage = () => {
    const newMessage = {
      timePeriod: new Date().toUTCString(),
      message,
      henchman,
    };
    let chatForHenchman = chat.get(henchman) || {draft: '', history: []};
    chatForHenchman = {
      history: [...chatForHenchman.history, newMessage],
      draft: '',
    };
    setChat(chat.set(henchman, chatForHenchman));
    setMessage('');
  };

  /**
   * sends message on enter.
   *
   * @param {Object} event
   */
  const sendMessageOnEnter = (event) => {
    if (event.key === 'Enter' && henchman && message.length > 0) {
      setMessage(event.target.value);
      sendMessage();
    }
  };

  /**
   * sets henchman name and focus on message input box.
   *
   * @param {string} name
   */
  const setHenchmanAndFocus = (name) => {
    const chatForHenchman = chat.get(name);
    const draft = chatForHenchman ? chatForHenchman.draft : '';
    setMessage(draft);
    setHenchman(name);
    inputEl.current.focus();
  };

  /**
   * get chat history for henchman.
   *
   * @returns {*}
   */

  const getChatHistory = () => {
    const chatForHenchman = chat.get(henchman);
    return chatForHenchman ? chatForHenchman.history : undefined;
  };

  /**
   * sets drafts message for the henchman.
   *
   * @param {Object} event
   */
  const setChatDraftForHenchman = (event) => {
    const chatForHenchman = chat.get(henchman) || {draft: '', history: []};
    chatForHenchman.draft = event.target.value;
    setChat(chat.set(henchman, chatForHenchman));
    setMessage(event.target.value);
  };

  return (
      <div className="app">
        <header className="app-header">
          <h4 className="app-title">Frog Bother Chat</h4>
          <h4>Vito Croakleone</h4>
        </header>
        <div className="chat-window" alt="chat-window">
          <HenchmenList setHenchman={(name) => setHenchmanAndFocus(name)} man={henchman}/>
          <ChatHistory chatHistory={getChatHistory()}/>
        </div>
        <div className="write-message">
          <TextField
              className="type-message"
              value={message}
              label="message-input"
              aria-labelledby="message-input"
              variant="outlined"
              onChange={(event) => setChatDraftForHenchman(event)}
              onKeyPress={(event) => sendMessageOnEnter(event)}
              inputRef={inputEl}
          />
          <Button
              disableElevation
              variant="contained"
              className="send-message"
              disabled={henchman.length === 0 || message.length === 0}
              onClick={(event) => sendMessage(event)}
          >
            Send
          </Button>
        </div>
      </div>
  );
};

export default App;
