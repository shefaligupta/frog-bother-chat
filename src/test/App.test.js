import React from 'react';
import renderer from 'react-test-renderer';
import {fireEvent, render} from '@testing-library/react';
import App from '../components/App';

it('should snapshot all the components', () => {
  const tree = renderer.create(<App/>).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should add chat history', () => {
  const {container} = render(<App/>);
  const textbox = container.querySelector('[type="text"]');
  fireEvent.change(textbox, {target: {value: 'this is a test message'}});
  fireEvent.click(container.querySelector('[type="button"]'));

  const chatHistory = container.querySelector('[class="chat-history"]');
  expect(chatHistory.children).toHaveLength(1);
});
