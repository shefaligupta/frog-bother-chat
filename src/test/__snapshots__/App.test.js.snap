// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should snapshot all the components 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <h4
      className="app-title"
    >
      Frog Bother Chat
    </h4>
    <h4>
      Vito Croakleone
    </h4>
  </header>
  <div
    alt="chat-window"
    className="chat-window"
  >
    <div
      className="henchman"
    >
      <div
        className="button-henchmen"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained active-button MuiButton-disableElevation"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Sonny
          </span>
        </button>
      </div>
      <div
        className="button-henchmen"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained select-henchman MuiButton-disableElevation"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Fredo
          </span>
        </button>
      </div>
      <div
        className="button-henchmen"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained select-henchman MuiButton-disableElevation"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Michael
          </span>
        </button>
      </div>
    </div>
    <div
      className="chat-history"
    >
      <div />
    </div>
  </div>
  <div
    className="write-message"
  >
    <div
      aria-labelledby="message-input"
      className="MuiFormControl-root MuiTextField-root type-message"
      onKeyPress={[Function]}
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink={false}
      >
        message-input
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-91 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-93"
          >
            <span>
              message-input
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained send-message MuiButton-disableElevation Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Send
      </span>
    </button>
  </div>
</div>
`;
